Add libvirt as a provider for openshift-ansible

lhuard1A opened the pull request:
`openshift-ansible` will be the tool to use to deploy OpenShift.
Today, it supports `GCE` and `AWS`.
This PR adds the `virt` provider that will use libvirt to create local VMs.

The idea is to be able to develop, test and debug applications, OpenShift and openshift-ansible locally on the developer’s workstation without to pay for cloud resources and without to suffer from potential network latency.

I tested those playbooks only with the new `bin/cluster` script (which required very little modifications). I haven’t tried to make them work with the older cluster.sh/cloud.rb scripts.

The commands that I have tested so far are:
```
bin/cluster create -m 1 -n 3 virt lenaic
```
It creates 4 local virtual machines with Qemu. An `osc get minions` on the master successfully reports 3 ready minions

```
bin/cluster terminate virt lenaic
```
Destroys the VMs
```

detiber commented on inventory/virt/group_vars/all:
```diff
@@ -0,0 +1,6 @@
+openshift_public_hostname: "{{ ansible_default_ipv4.address }}"
+openshift_ip: "{{ ansible_default_ipv4.address }}"
+openshift_public_ip: "{{ ansible_default_ipv4.address }}"
+openshift_env: "{{ oo_env }}"
```
You shouldn't need to define openshift_hostname, openshift_public_hostname, openshift_ip, or openshift_public_ip, the openshift_facts module should default these correctly.

openshift_env and oo_env vars are no longer used.

detiber commented on inventory/virt/group_vars/tag_host-type-master:
```diff
@@ -0,0 +1,4 @@
+openshift_api_url: https://{{ openshift_hostname }}:8443
+openshift_api_public_url: https://{{ openshift_public_hostname }}:8443
+openshift_webui_url: https://{{ openshift_hostname }}:8444
+openshift_webui_public_url: https://{{ openshift_public_hostname }}:8444
```
These vars shouldn't default correctly without needing to set them.

detiber commented on inventory/virt/group_vars/tag_host-type-node:
```diff
@@ -0,0 +1,5 @@
+openshift_node_memory:
+openshift_node_pod_cidr:
+openshift_node_labels: {}
+openshift_node_annotations: {}
```
These vars should default correctly without needing to set them.

detiber commented on inventory/virt/group_vars/tag_host-type-openshift-master:
```diff
@@ -0,0 +1 @@
+tag_host-type-master
```
This symlink shouldn't be needed, since the above vars aren't needed.

detiber commented on inventory/virt/group_vars/tag_host-type-openshift-node:
```diff
@@ -0,0 +1 @@
+tag_host-type-node
```
This symlink shouldn't be needed, since the above vars aren't needed.

detiber commented on playbooks/virt/openshift-cluster/launch_instances.yml:
```diff
@@ -0,0 +1,102 @@
+  get_url:
+    url: http://download.fedoraproject.org/pub/fedora/linux/releases/21/Cloud/Images/x86_64/Fedora-Cloud-Base-20141203-21.x86_64.qcow2
+    sha256sum: 3a99bb89f33e3d4ee826c8160053cdb8a72c80cd23350b776ce73cd244467d86
+    dest: '{{ libvirt_storage_pool_path }}/Fedora-Cloud-Base-20141203-21.x86_64.qcow2'
```
I think I'd like to see this default to CentOS and be overridable through variables.

twiest replied:
I agree with this

lhuard1A replied:
Does CentOS also release “cloud images”? Either in qcow2 or even in raw format?

On the [CentOS download page](https://www.centos.org/download/), I could only find ISOs of LiveDVDs, Amazon’s AMIs and docker base image.
Nothing as convenient to use as a “cloud image”.

detiber commented on playbooks/virt/openshift-cluster/meta-data:
```diff
@@ -0,0 +1,2 @@
+instance-id: {{ item[0] }}
+local-hostname: {{ item[0] }}
```
Could you create a templates directory and put the meta-data, user-data and domain.xml files there?

detiber commented on bin/cluster:
```diff
@@ -132,7 +134,7 @@ if __name__ == '__main__':
     cluster = Cluster()
 
-    providers = ['gce', 'aws']
+    providers = ['gce', 'aws', 'virt']
```
I'd like to see this named libvirt, since virt is a bit too vague.

twiest replied:
Heh, sorry, I didn't read @detiber 's comments before posting mine. I agree completely. :)

detiber commented:
Overall, I love where this is going.

@twiest I'd like to see if we can get this reviewed, updated and merged before https://github.com/openshift/openshift-ansible/pull/139 So that @lhuard1A does not have to deal with a major refactor.

twiest commented on README_virt.md:
```diff
@@ -0,0 +1,92 @@
+VIRT Setup instructions
+=======================
+
+`virt` is an `openshift-ansible` provider that uses `libvirt` to create local Fedora VMs that are provisioned exactly the same way that cloud VMs would be provisioned.
```
Given this, should we maybe call this "libvirt" instead of just "virt" ?

I can see us adding other virt providers in the future, and "virt" will make this confusing.

So, README_libvirt.md

twiest commented on bin/cluster:
```diff
@@ -87,6 +87,8 @@ class Cluster(object):
                 os.environ[key] = config.get('ec2', key)
 
             inventory = '-i inventory/aws/ec2.py'
+        elif 'virt' == provider:
```
same, change to libvirt

twiest commented:
@lhuard1A this is really cool stuff, thanks for adding it. I think it'll be ready to merge once the comments have been addressed.

lhuard1A commented:
@detiber, @twiest: thanks very much for your quick feedback.
I have rebased, renamed `virt` to `libvirt`, removed group_vars, put templates in a dedicated subdir.

Concerning the base image used by the VMs, would you know where a “cloud image” of CentOS could be downloaded? Ideally in a qcow2 format.
I couldn’t find any on the [CentOS download page](https://www.centos.org/download/).
The [Atomic images](https://wiki.centos.org/SpecialInterestGroup/Atomic/Download) are not suitable as Atomic images are not intended to be managed classically by yum.

lhuard1A commented:
When I initially tested this PR, it worked perfectly. It is now failing at the [openshift_register_nodes | Create node client certificates] step because `openshift admin` doesn’t implement the `create-node-cert` command anymore.
As this comes from a common role, I don’t think that this error is linked to that PR.
Here is where that command disappeared: [origin/28e7369](https://github.com/openshift/origin/commit/28e7369457cc5abfb33ac801a43967f03956d3b4).

twiest commented:
@lhuard1A this looks really good to me.

If @detiber confirms that the error you're hitting is a problem overall (not created by this PR), then I think this is probably ready to merge.

detiber commented:
@lhuard1A indeed, I have fixes for the other providers lined up along with a massive refactor in: https://github.com/openshift/openshift-ansible/pull/139

As far as the CentOS images, they have generic cloud images available here: http://cloud.centos.org/centos/7/images/, but I can take care of that when I rebase my PR (I'll probably go ahead and add a flag to choose between Fedora and CentOS so that we can run tests against both). I'm actually quite surprised that we didn't run into more issues running on Fedora 21, since we've been testing only on RHEL so far.

@twiest, lets get this merged and I'll work on getting the necessary updates in place.  @nhr will be happy to start playing with this as well.

lhuard1A commented:
Thanks @detiber for the link. I’ve added a commit to use a CentOS image instead of a Fedora one.

With CentOS, I reach the same error as with Fedora because `openshift admin` doesn’t know the `create-node-cert` command anymore.
The fix seems to be there in #139 : https://github.com/openshift/openshift-ansible/pull/139/files#diff-d58243d309d7ab3a5f8254c9505f9703 (`create-node-cert` is replaced by `create-node-config`).

Please let me know when you want me to squash the commits of that PR.

twiest commented:
@lhuard1A ok, I think we're close to ready to merge. Do whatever you need and then let me know and I'll merge it.

Thanks! :)

detiber commented:
@lhuard1A Feel free to squash what you have now, then we can merge and I can rebase my branch (I'd like to avoid having to force you to rebase your work on top of mine, since it adds the common playbooks and the bin/cluster changes for supporting origin and evenutally enterprise).

lhuard1A commented:
@twiest : I have squashed the commits. You can merge it now.
Thanks !

twiest commented:
:+1:
